Update AltUseVendors

Siggless opened the pull request:
Added buy/sell sound effects and voice lines.

Siggless committed:
Update AltUseVendors

Added buy/sell sound effects and voice lines.

apple1417 commented on AltUseVendors/__init__.py:
```diff
@@ -73,6 +73,29 @@ class ShopInfo:
+    Game.BL2: {
+        "InteractiveObj_VendingMachine_GrenadesAndAmmo": "Ake_VOCT_Contextual.Ak_Play_VOCT_Marcus_Vending_Munition_Purchase",
+        "InteractiveObj_VendingMachine_HealthItems": "Ake_VOCT_Contextual.Ak_Play_VOCT_Zed_Store_Welcome",
+        #"VendingMachine_Weapons_Definition": "Ake_VOCT_Contextual.Ak_Play_VOCT_Marcus_Vending_Munition_Bye",
+        # DLC Vendors
+        "IO_Aster_VendingMachine_GrenadesAndAmmo": "Ake_Aster_VO.VOCT.Ak_Play_VOCT_Aster_Marcus_Store_Purchase",
+        "IO_Aster_VendingMachine_HealthItems": "Ake_Aster_VO.VOCT.Ak_Play_VOCT_Aster_Zed_Store_Purchase",
+        #"VendingMachine_Aster_Weapons_Definition": "Ake_Aster_VO.VOCT.Ak_Play_VOCT_Aster_Marcus_Store_Bye",
+        #"VendingMachine_TorgueToken": "Ake_Iris_VO.Ak_Play_Iris_TorgueVendingMachine_Purchase",
```
What's up with the commented out ones?

Siggless replied:
Those voiceline didn't make sense as you were selling items. I left them in to show the vendor names but I've removed them in the new commit.

apple1417 replied:
After playing with it a bit in game, I disagree, I think they generally fit ok, and it's better to have a line on every vendor instead of only some

apple1417 commented on AltUseVendors/__init__.py:
```diff
@@ -73,6 +73,29 @@ class ShopInfo:
+        "InteractiveObj_VendingMachine_HealthItems": "Ake_Cork_VOCT_Contextuals.Cork_VOCT_NurseNina.Ak_Play_VOCT_Cork_NurseNina_Store_Purchase",
+    },
+    Game.AoDK: {
+        # Don't know other game AkEvent names
```
open blcmm has aodk dumps, at a guess it'll just be the aster ones

Siggless replied:
Found them and also added in TPS DLC ones.

apple1417 commented on AltUseVendors/__init__.py:
```diff
@@ -106,6 +129,7 @@ def sell_trash(PC: WillowPlayerController, vendor: WillowVendingMachine) -> None
         vendor: The vendor they're selling at.
     """
     PC.GetPawnInventoryManager().SellAllTrash()
+    PC.pawn.PlayAkEvent(unrealsdk.FindObject("AkEvent", AKE_SELL))
```
`PC.Pawn` - copy the capitalization when using unreal property names.

Siggless replied:
Done

apple1417 commented on AltUseVendors/__init__.py:
```diff
@@ -378,6 +404,10 @@ def PerformedSecondaryUseAction(self, caller: unrealsdk.UObject, function: unrea
         info.purchase_function(caller, vendor)
 
+        # Play the dialogue sound if one is given for this vendor
+        if vendor.InteractiveObjectDefinition.name in AKE_DICT[Game.GetCurrent()]:
+            vendor.PlayAkEvent(unrealsdk.FindObject("AkEvent", AKE_DICT[Game.GetCurrent()][vendor.InteractiveObjectDefinition.name]))
```
as above, `vendor.InteractiveObjectDefinition.Name`

you can save doing this dict lookup twice using something like
```py
event = AKE_DICT[Game.GetCurrent()].get(vendor.InteractiveObjectDefinition.Name)
if event is not None:
  ...
```

Siggless replied:
Done

apple1417 commented on AltUseVendors/__init__.py:
```diff
@@ -73,6 +73,29 @@ class ShopInfo:
 
 
+# Sound AkEvents for playing
+AKE_BUY: str = "Ake_UI.UI_Vending.Ak_Play_UI_Vending_Buy"
+AKE_SELL: str = "Ake_UI.UI_Vending.Ak_Play_UI_Vending_Sell"
+AKE_DICT = {
```
Buy and sell are reasonable enough names, but what's 'dict' supposed to mean? Prefer something like `AKE_PURCHASE_BY_VENDOR`.

You could do the game lookup here once at the root level, rather than needing to do so each time, since it'll never change.
```py
AKE_PURCHASE_BY_VENDOR = {
  Game.BL2: ...
}[Game.GetCurrent()]
```

Siggless replied:
Changed to AKE_VOICELINE_BY_VENDOR and looked up the game like you have there.

apple1417 commented on AltUseVendors/__init__.py:
```diff
@@ -183,6 +207,7 @@ def refill_ammo(PC: WillowPlayerController, vendor: WillowVendingMachine) -> Non
     """
     for _, _, pool in iter_ammo_data(PC, vendor):
         pool.SetCurrentValue(pool.GetMaxValue())
+    PC.pawn.PlayAkEvent(unrealsdk.FindObject("AkEvent", AKE_BUY))
```
Alt use vendors has a bit of an undeserved reputation for causing stutters, despite having been optimized several times. Find object is a somewhat expensive call, so I want to be careful about adding it. It may well turn out to be the best solution, but there's some things I want checked first.

Firstly, I assume these objects are map specific, so we can't do a root level find object? And that they depend on other assets so we can't construct a new version like we do interaction icons? If we can get a single global reference, prefer that.

Now if that's not possible (which I'm pretty sure it isn't), can we use keep alive. On first use, do use a find object, but then call keep alive, and keep the direct object reference for all future calls.

Regardless of what solution ends up best, I also want to see some handling for if find object returns None. If it's a map specific object, there's always a chance of something weird happening on some map you didn't test, don't want to cause errors unwittingly.

Siggless replied:
I don't 100% understand how loading objects and storing references to them works, but using FindObject outside the class or on mod Enable returns None.
Added `find_and_play_akevent` method to handle storing and retrieving them when called.

Siggless committed:
Update AltUseVendors

Pull request fixes.
Added find_and_play_akevent method so that we only FindObject the first time we play an event.

Siggless reviewed (commented):
Comments for new commit.

apple1417 commented on AltUseVendors/__init__.py:
```diff
@@ -76,25 +76,41 @@ class ShopInfo:
     }
-}
+}[Game.GetCurrent()]
+LOADED_AKEVENTS = {}
```
This is not constant, so should not be in all caps.

apple1417 commented on AltUseVendors/__init__.py:
```diff
@@ -76,25 +76,41 @@ class ShopInfo:
+}[Game.GetCurrent()]
+LOADED_AKEVENTS = {}
+
+def find_and_play_akevent(caller:unrealsdk.UObject, eventName: str) -> None:
```
`caller:unrealsdk.UObject` - caller is a bit of a useless name, it's convention for use in hooks but doesn't tell you anything here You're also not actually passing the caller, you're actually passing in a `pawn: WillowPlayerPawn` - though everywhere else in the mod uses a controller, so I changed it to that instead.

`eventName` - follow pep8, use camel case `event_name`

apple1417 commented on AltUseVendors/__init__.py:
```diff
@@ -76,25 +76,41 @@ class ShopInfo:
+    """
+    Finds an AkEvent object if not already stored in LOADED_AKEVENTS, then plays it.
+    """
+    if eventName is not None:
```
You typed `eventName` as `str`, so it can never possibly be none

apple1417 commented on AltUseVendors/__init__.py:
```diff
@@ -76,25 +76,41 @@ class ShopInfo:
+    if eventName is not None:
+        event = LOADED_AKEVENTS.get(eventName)
+        if event is None:
+            event = unrealsdk.FindObject("AkEvent", eventName)
```
This may still return none sometimes, and we don't want to try cache/keep alive that.

apple1417 commented on AltUseVendors/__init__.py:
```diff
@@ -405,8 +421,8 @@ def PerformedSecondaryUseAction(self, caller: unrealsdk.UObject, function: unrea
         # Play the dialogue sound if one is given for this vendor
-        if vendor.InteractiveObjectDefinition.name in AKE_DICT[Game.GetCurrent()]:
-            vendor.PlayAkEvent(unrealsdk.FindObject("AkEvent", AKE_DICT[Game.GetCurrent()][vendor.InteractiveObjectDefinition.name]))
+        dialogEvent = AKE_VOICELINE_BY_VENDOR.get(vendor.InteractiveObjectDefinition.Name)
+        find_and_play_akevent(vendor, dialogEvent)
```
`dialogEvent` (pep8) may be None, but you defined the second arg to `find_and_play_akevent` as just `str`, so you're not allowed to pass that. Need an explicit check for it here - which we can also use to print a debug message in case we haven't covered all vendor types.

apple1417 reviewed (commented):
I did these fixes myself, but figured I'd still give a review so you know the justifications behind the changes.

The other thing I'd have to ask if I hadn't done it is bumping the version number, updating the changelog, and updating the zip
